Wikipedia/Code/WebViewController.m

```diff
@@ -86,121 +86,154 @@ - (instancetype)initWithSession:(SessionSingleton*)aSession {
-                                                              dataFileWidth:imageClicked[@"data-file-width"]
-                                                             dataFileHeight:imageClicked[@"data-file-height"]];
-            
-            if (![imageTagClicked isSizeLargeEnoughForGalleryInclusion]) {
-                return;
+
+            else if (body[@"imageClicked"]) {
+                NSDictionary* imageClicked = [body[@"imageClicked"] wmf_dictionaryByRemovingNullObjects];
+
+                WMFImageTag* imageTagClicked = [[WMFImageTag alloc] initWithSrc:imageClicked[@"src"]
+                                                                         srcset:nil
+                                                                            alt:nil
+                                                                          width:imageClicked[@"width"]
+                                                                         height:imageClicked[@"height"]
+                                                                  dataFileWidth:imageClicked[@"data-file-width"]
+                                                                 dataFileHeight:imageClicked[@"data-file-height"]];
+
+                if (![imageTagClicked isSizeLargeEnoughForGalleryInclusion]) {
+                    return;
+                }
+
+                NSString* selectedImageSrcURLString = imageClicked[@"src"];
+                NSCParameterAssert(selectedImageSrcURLString.length);
+                if (!selectedImageSrcURLString.length) {
+                    DDLogError(@"Image clicked callback invoked with empty src url: %@", imageClicked);
+                    return;
+                }
+
+                NSURL* selectedImageURL = [NSURL URLWithString:selectedImageSrcURLString];
+
+                selectedImageURL = [selectedImageURL wmf_imageProxyOriginalSrcURL];
+
+                [self.delegate webViewController:self didTapImageWithSourceURL:selectedImageURL];
+            } else if (body[@"referenceClicked"]) {
+                [self referencesShow:body[@"referenceClicked"]];
+            } else if (body[@"editClicked"]) {
+                NSUInteger sectionIndex = (NSUInteger)[body[@"editClicked"][@"sectionId"] integerValue];
+                if (sectionIndex < [self.article.sections count]) {
+                    [self.delegate webViewController:self didTapEditForSection:self.article.sections[sectionIndex]];
+                }
+            } else if (body[@"nonAnchorTouchEndedWithoutDragging"]) {
+                [self referencesHide];
             }
-            
-            NSString* selectedImageSrcURLString = imageClicked[@"src"];
-            NSCParameterAssert(selectedImageSrcURLString.length);
-            if (!selectedImageSrcURLString.length) {
-                DDLogError(@"Image clicked callback invoked with empty src url: %@", imageClicked);
-                return;
+        }
+    }
+
+    if ([message.body isKindOfClass:[NSString class]]) {
```

Comment:
The message "name" checks below should, I think, be part of the message name "if" check above. Formerly the single "if" above was basically functioning like to a "switch" for all the name of messages which could be received. But here we've broken out a second "name if" structure, and inverted the logic, now checking the body before the name.